src/Bicep.Core/TypeSystem/ResourceTypeCache.cs

```diff
@@ -1,34 +1,22 @@
 // Copyright (c) Microsoft Corporation.
 // Licensed under the MIT License.
 using System;
-using System.Collections.Generic;
 using Bicep.Core.Resources;
 using System.Collections.Concurrent;
 
 namespace Bicep.Core.TypeSystem
 {
     public class ResourceTypeCache
     {
-        private class KeyComparer : IEqualityComparer<(ResourceTypeGenerationFlags flags, ResourceTypeReference type)>
-        {
-            public static IEqualityComparer<(ResourceTypeGenerationFlags flags, ResourceTypeReference type)> Instance { get; }
-                = new KeyComparer();
-
-            public bool Equals((ResourceTypeGenerationFlags flags, ResourceTypeReference type) x, (ResourceTypeGenerationFlags flags, ResourceTypeReference type) y)
-                => x.flags == y.flags &&
-                    ResourceTypeReferenceComparer.Instance.Equals(x.type, y.type);
-
-            public int GetHashCode((ResourceTypeGenerationFlags flags, ResourceTypeReference type) x)
-                => x.flags.GetHashCode() ^
-                    ResourceTypeReferenceComparer.Instance.GetHashCode(x.type);
-        }
+        private record CacheKey(
+            ResourceTypeGenerationFlags Flags,
+            ResourceTypeReference Type);
 
-        private readonly ConcurrentDictionary<(ResourceTypeGenerationFlags flags, ResourceTypeReference type), ResourceTypeComponents> cache
-            = new ConcurrentDictionary<(ResourceTypeGenerationFlags flags, ResourceTypeReference type), ResourceTypeComponents>(KeyComparer.Instance);
+        private readonly ConcurrentDictionary<CacheKey, ResourceTypeComponents> cache = new();
```

Comment:
I think the `CacheKey` record can be replaced with a value tuple `(ResourceTypeGenerationFlags, ResourceTypeReference)`.

Reply:
Is there a great benefit either way? I just used a record to reduce clutter.

Reply:
Just realized that that's the original implementation. Any reason why we replace the value tuple with a record?

Reply:
Ah we commented at the same time. Value tuple is a value type so it could avoid heap allocations.

Reply:
In that case, `record struct` feels like a good compromise!

Reply:
Yeah, that definitely works!